### PR TITLE
WSTEMA 1329 Minimum Scope: Remove media player component, and add MediaLoader instead to OD TV pages

### PR DESCRIPTION
**Pull request by holchris**
Resolves JIRA [1329](https://jira.dev.bbc.co.uk/browse/WSTEAMA-1329)
Related to https://github.com/bbc/simorgh/pull/11978

Overall changes
======

Code changes
======
- Replaces AVPlayer with MediaLoader component

Testing
======
- adds unit test for Hausa On Demand TV Episode

### CODE REVIEW

File: src/integration/pages/onDemandTVPage/somali/canonical.test.js
Comment (karinathomasbbc) [inline src/integration/pages/onDemandTVPage/somali/canonical.test.js:1]: Reminder: we need to replace the somali (brand) test with a hausa (episode) one, to make sure that the page works for both TV Programmes & TV Episodes.

**Review by karinathomasbbc — APPROVED** [review]
Wohoooooo! Nice work!

**Comment by amoore108** [comment]
Looks good 👍 

I get an SMPErr when hitting http://localhost.bbc.com:7080/hausa/bbc_hausa_tv/tv/w172xcg0kg6vph8

Is that expected or do I need to add in any renderer_env stuff?

**Comment by holchris** [comment]
> Looks good 👍
> 
> I get an SMPErr when hitting http://localhost.bbc.com:7080/hausa/bbc_hausa_tv/tv/w172xcg0kg6vph8
> 
> Is that expected or do I need to add in any renderer_env stuff?

When visiting the live page for http://bbc.com/hausa/bbc_hausa_tv/tv/w172xcg0kg6vph8 the media player doesn't load and displays the error message _This program has now been discontinued._ (translated curtesy of google!) 

The local fixture data for this page is outdated, which is why the error you saw is getting returned! To check this I updated the local fixture data to match live and now when visiting local host it displays the 
_This program has now been discontinued._ instead of the SMPErr error you saw. 

TLDR: out of data local fixture data is the cause of the SMPErr

**Comment by amoore108** [comment]
> TLDR: out of data local fixture data is the cause of the SMPErr

Ah I see, thanks for clarifying! 👍

**Review by amoore108 — APPROVED** [review]

File: src/app/components/MediaLoader/index.tsx [hunk 1]
```diff
@@ -145,6 +145,7 @@ type Props = {
   blocks: MediaBlock[];
   className?: string;
   embedded?: boolean;
+  episodePageIdentifier: string;
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.tsx]: We need to use a variable name which is more generic, because this media loader should not need to know about an episode page identifier. 

However, naming things is the hardest thing in the world. 

Given that it's the counterName that we're trying replace with this variable, I am thinking something like `counterNameOverride` but don't love that. Let's see if we can find something more suitable.

File: src/app/pages/OnDemandTvPage/OnDemandTvPage.tsx [hunk 2]
```diff
@@ -108,17 +107,13 @@ const OnDemandTvPage = ({
...
-    ['mediaAssetPage', 'videoPlayer'],
-    translations,
-  );
-
   const hasRecentEpisodes = recentEpisodes && Boolean(recentEpisodes.length);
   const metadataTitle = episodeTitle
     ? `${brandTitle} - ${episodeTitle} - ${brandName}`
     : headline;
 
+  const episodePageIdentifier = `${service}.bbc_${service}_tv.tv.${episodeId}.page`;
```
Comment (karinathomasbbc) [inline src/app/pages/OnDemandTvPage/OnDemandTvPage.tsx]: Let's definitely add a test for this - we can check that the MediaLoader is called with this variable (can help you figure out how to so that).

File: src/app/components/MediaLoader/index.tsx (hunk 1 above)
Comment (holchris) [reply src/app/components/MediaLoader/index.tsx]: I will have a think of a variable name!

Comment (holchris) [reply src/app/components/MediaLoader/index.tsx]: Do we want this variable to be optional?

File: src/app/legacy/containers/PageHandlers/withContexts.jsx [hunk 3]
```diff
@@ -60,6 +60,7 @@ const WithContexts = Component => {
             mvtExperiments={mvtExperiments}
             isNextJs={isNextJs}
             isUK={isUK}
+            counterName={atiAnalytics?.pageIdentifier ?? null}
```
Comment (karinathomasbbc) [inline src/app/legacy/containers/PageHandlers/withContexts.jsx]: This is no longer required, and needs to be deleted: see https://github.com/bbc/simorgh/pull/11782/files#r1750005785
```suggestion
```

File: src/app/components/MediaLoader/index.tsx [hunk 4]
```diff
@@ -144,10 +144,16 @@ const MediaContainer = ({ playerConfig, showAds }: MediaContainerProps) => {
 type Props = {
   blocks: MediaBlock[];
   className?: string;
+  counterNameOverride?: string;
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.tsx]: This should now be `pageIdentifierOverride` instead of countername override: see https://github.com/bbc/simorgh/pull/11782/files#r1750005785
```suggestion
  pageIdentifierOverride?: string;
```

**Review by karinathomasbbc — CHANGES_REQUESTED** [review]

**Comment by karinathomasbbc** [comment]
This is dependent on #11782 being merged first

File: src/app/components/MediaLoader/index.test.tsx [hunk 5]
```diff
@@ -168,4 +169,25 @@ describe('MediaLoader', () => {
       });
     });
   });
+
+  describe('Config', () => {
+    it('should use the counterNameOverride when rendering On Demand TV', async () => {
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.test.tsx]:
```suggestion
    it('should use the pageIdentifierOverride when rendering On Demand TV', async () => {
```

File: src/app/components/MediaLoader/index.tsx [hunk 6]
```diff
@@ -169,7 +175,7 @@ const MediaLoader = ({ blocks, embedded, className }: Props) => {
   const config = buildConfig({
     id,
     blocks,
-    counterName,
+    counterName: pageIdentifierOverride || counterName,
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.tsx]: This might cause a conflict with Aaron's PR, but whoever merges first will have to fix it!

File: src/app/components/MediaLoader/index.test.tsx (hunk 5 above)
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.test.tsx]: Do we need an opposite / positive test for this, whereby it uses the counter name from the RequestContext if no override is passed in?

**Comment by karinathomasbbc** [comment]
I have also realised that we have not touched the Cypress tests for the On Demand TV page, however, it still refers to the iFrame: https://github.com/bbc/simorgh/blob/4b42906e8f86f3666c435885db8baad01b756aba/cypress/e2e/pages/onDemandTV/tests.js#L22

I believe these tests will fail when we run them against the test environment (just to confirm, run `yarn test:e2e:interactive`, and select the On Demand TV page)

We will need to update these tests to reflect the new media loader functionality, and looks like we need to refactor these tests (from NextJS Cypress tests: https://github.com/bbc/simorgh/blob/c9d8565bb51dbb90c40e0c2a7aaec6adf1608181/ws-nextjs-app/cypress/e2e/livePage/mediaPlayer.ts

File: cypress/support/config/settings.js [hunk 7]
```diff
@@ -4223,7 +4223,7 @@ module.exports = () => ({
               '/pashto/bbc_pashto_tv/tv_programmes/w13xttn4', // Brand
               '/pashto/bbc_pashto_tv/tv/w172xcldhhrhmcf', // Episode
             ],
-            enabled: false,
+            enabled: true,
```
Comment (karinathomasbbc) [inline cypress/support/config/settings.js:4226]: We may need to revert this - let's fix the cypress tests first, and then see whether the pashto tests fail miserably or not.

File: cypress/support/config/settings.js [hunk 8]
```diff
@@ -2231,7 +2231,7 @@ module.exports = () => ({
             enabled: true,
           },
         },
-        smoke: false,
+        smoke: true,
```
Comment (karinathomasbbc) [inline cypress/support/config/settings.js:2234]: This enables smoke tests for hindi On Demand TV on local (which will fail, as the tests are expecting an iFrame)

File: src/app/components/MediaLoader/index.test.tsx [hunk 9]
```diff
@@ -168,4 +169,25 @@ describe('MediaLoader', () => {
...
+      const buildConfigSpy = jest.spyOn(buildConfig, 'default');
+      await act(async () => {
+        render(
+          <MediaPlayer
+            blocks={onDemandTvBlocks as MediaBlock[]}
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.test.tsx]: Did we want to import the json from https://github.com/bbc/simorgh/blob/84335ed35db5ef6d240b426a6294a0aa3902c260/data/hindi/bbc_hindi_tv/tv_programmes/w13xttlw.json instead of duplicating the fixture data in src/app/components/MediaLoader/fixture.ts?

File: cypress/support/config/settings.js (hunk 7 above)
Comment (karinathomasbbc) [reply cypress/support/config/settings.js:4226]: I can see that the local pashto episode test `has valid iframe` is passing which we are not expecting. I think we should not run the pashto E2E tests in local, as this is covered by integration tests.

Comment (karinathomasbbc) [reply cypress/support/config/settings.js:4226]: We determined that the pashto episode tests were working OK due to some logic in the tests which checks if the episode is available or not. We can keep this test on local.

File: src/app/components/MediaLoader/utils/buildSettings.ts [hunk 10]
```diff
@@ -38,6 +43,7 @@ const buildSettings = ({
     enableToucan: true,
     appType: isAmp ? 'amp' : 'responsive',
     appName: service !== 'news' ? `news-${service}` : 'news',
+    externalEmbedUrl: '',
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/utils/buildSettings.ts]: This should actually be removed, as On Demand TV pages cannot be embedded. The reason the test was failing was due to a merge conflict being incorrectly resolved (my bad oops), so let's remove this from here, and remove `externalEmbedUrl: ''` from the test as well.

File: src/app/components/MediaLoader/utils/buildSetting.test.ts [hunk 11]
```diff
@@ -613,7 +483,7 @@ describe('buildSettings', () => {
           enableToucan: true,
           appType: 'responsive',
           autoplay: false,
-          mediator: { host: 'open.test.bbc.co.uk' },
+          externalEmbedUrl: '',
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/utils/buildSetting.test.ts]: This should be deleted
```suggestion
```

File: src/app/components/MediaLoader/utils/buildSettings.ts (hunk 10 above)
Comment (Nabeel1276) [reply src/app/components/MediaLoader/utils/buildSettings.ts]: Ahh, I see. Thanks for letting me know.

File: cypress/e2e/pages/onDemandTV/tests.js [hunk 12]
```diff
@@ -19,7 +16,7 @@ export default ({ service, pageType, variant }) => {
         retries: 3,
       },
       () => {
-        it('should render an iframe with a valid URL', () => {
+        it('should render a valid media player', () => {
```
Comment (karinathomasbbc) [inline cypress/e2e/pages/onDemandTV/tests.js:19]: While we're in here, could we use cy.getPageData instead of cy.request?

File: cypress/e2e/pages/onDemandTV/tests.js [hunk 13]
```diff
@@ -29,34 +26,15 @@ export default ({ service, pageType, variant }) => {
...
-                type: 'text/html',
-                allowFallback: true,
+            cy.get('[data-e2e="media-loader__container"]').should('be.visible');
+            cy.get('[data-e2e="media-loader__placeholder"] div')
+              .eq(1)
+              .within(() => {
+                cy.get('img')
+                  .should('be.visible')
+                  .should('have.attr', 'src')
+                  .should('not.be.empty');
```
Comment (karinathomasbbc) [inline cypress/e2e/pages/onDemandTV/tests.js]: Should we check that the img source matches what's in the data, or is that overkill?

File: src/app/pages/OnDemandTvPage/OnDemandTvPage.tsx [hunk 14]
```diff
@@ -174,18 +169,10 @@ const OnDemandTvPage = ({
...
-              title="On-demand TV"
-              iframeTitle={iframeTitle}
-              hasBottomPadding={false}
-              skin="classic"
-              showLoadingImage
-              darkPlaceholder
+            <MediaLoader
+              blocks={pageData?.mediaBlocks}
               css={styles.mediaPlayer}
+              pageIdentifierOverride={episodePageIdentifier}
```
Comment (karinathomasbbc) [inline src/app/pages/OnDemandTvPage/OnDemandTvPage.tsx]: Can we add a test to the OnDemandTV test suite, to ensure that when we render the OD TV page, then the MediaLoader is called with the correct pageIdentifierOverride? We do have a similar test in the Media Loader, but I would like a test at this level too, just to make sure.

File: cypress/e2e/pages/onDemandTV/tests.js (hunk 12 above)
Comment (alex-magana) [reply cypress/e2e/pages/onDemandTV/tests.js:19]: Looking at the implementation [here](https://github.com/bbc/simorgh/blob/9c1b673a00f789c73d822e89263e1dd692052bbc/cypress/support/commands/application.js#L128), `getPageData` makes a call to the BFF for non-local environments.
Since we don't support On Demand TV in the BFF, would it be appropriate to invoke `getPageData` here?
Also, given the recent work to remove dependence on the BFF for the e2es, would
invoking `getPageData` here be a reinstatement of the said dependency?

Comment (karinathomasbbc) [reply cypress/e2e/pages/onDemandTV/tests.js:19]: Oops, I actually meant the `cy.getPageDataFromWindow` which gets it from window.SIMORGH_DATA, which only works on canonical, but that's OK because we don't support AMP on these pages.

File: cypress/support/config/settings.js (hunk 8 above)
Comment (alex-magana) [reply cypress/support/config/settings.js:2234]: Marking this as resolved on account of the tests having been updated to
assert for existence of the MediaLoader component and the placeholder image.

File: src/app/components/MediaLoader/index.test.tsx (hunk 9 above)
Comment (alex-magana) [reply src/app/components/MediaLoader/index.test.tsx]: As these are unit tests, we've sought to replicate the object created [here](https://github.com/bbc/simorgh/blob/1c1871222b1f47311d68123cfbc9d0ec50da07ab/src/app/routes/onDemandTV/getInitialData/index.js#L42) and passed down to the MediaLoader.
This is the rationale behind creating [this](https://github.com/bbc/simorgh/blob/1c1871222b1f47311d68123cfbc9d0ec50da07ab/src/app/components/MediaLoader/fixture.ts#L275) fixture.
On account of this, is it okay to retain `onDemandTvBlocks` in the test?

File: cypress/e2e/pages/onDemandTV/tests.js (hunk 12 above)
Comment (alex-magana) [reply cypress/e2e/pages/onDemandTV/tests.js:19]: Ah! Got it.
Thanks for the clarification. Let me give this a go.

File: src/app/components/MediaLoader/index.test.tsx (hunk 9 above)
Comment (karinathomasbbc) [reply src/app/components/MediaLoader/index.test.tsx]: Sure, it's very unlikely to change in the future, keeps our tests stable, and is consistent with the other tests in this file.

File: src/app/components/MediaLoader/index.test.tsx [hunk 15]
```diff
@@ -168,4 +169,51 @@ describe('MediaLoader', () => {
...
+    it('should use the pageIdentifier from the EventTrackingContext', async () => {
+      const buildConfigSpy = jest.spyOn(buildConfig, 'default');
+      await act(async () => {
+        render(
+          <MediaPlayer blocks={onDemandTvBlocks as MediaBlock[]} embedded />,
+          {
+            service: 'hindi',
+            pageData: {
+              mediaBlocks: onDemandTvBlocks,
+              pageIdentifier: 'hindi.bbc_hindi_tv.tv_programmes.w13xttlw.page',
```
Comment (karinathomasbbc) [inline src/app/components/MediaLoader/index.test.tsx:236]: I think the `pageIdentifier` should be set on the `atiData` rather than the `pageData` object and passed in as a param to the renderer. https://github.com/bbc/simorgh/blob/bb2b3a935f02ba359eb065cd3349bcfa093e9631/src/app/components/ATIAnalytics/params/genericPage/buildParams.ts#L23

Comment (alex-magana) [reply src/app/components/MediaLoader/index.test.tsx:236]: That isn't the case I'm afraid. 
`atiData` is used for pages whose analytics business logic was migrated to the BFF.
See [isMigrated](https://github.com/bbc/simorgh/blob/9ff521be38586ccbd7488c9d770664e5be3a1937/src/app/components/ATIAnalytics/params/index.ts#L159).
For On Demand TV, we rely on [buildTvRadioATIParams](https://github.com/bbc/simorgh/blob/9ff521be38586ccbd7488c9d770664e5be3a1937/src/app/components/ATIAnalytics/params/tvRadioPage/buildParams.ts#L16). This function retrieves `pageIdentifier` from pageData which is in turn set [here](https://github.com/bbc/simorgh/blob/8277e903fdcd3f5d32fc0d753d100516dc9f53d4/src/app/routes/onDemandTV/getInitialData/index.js#L72).

File: cypress/e2e/pages/onDemandTV/tests.js (hunk 13 above)
Comment (alex-magana) [reply cypress/e2e/pages/onDemandTV/tests.js]: We should be able to do this. We'll check how to extract the path from the pageData
and add an assertion to this end.

File: src/app/components/MediaLoader/index.test.tsx (hunk 15 above)
Comment (karinathomasbbc) [reply src/app/components/MediaLoader/index.test.tsx:236]: My apologies! Thanks for figuring that out, I only had a quick look and totally forgot about this mish-mash of migrated and un-migrated page types.

**Review by karinathomasbbc — APPROVED** [review]
Amazing work, and thank you so much for addressing all of my feedback! 

🚢 🚢 🚢 🚢 🚢 🚢

File: src/app/pages/OnDemandTvPage/OnDemandTvPage.tsx (hunk 14 above)
Comment (alex-magana) [reply src/app/pages/OnDemandTvPage/OnDemandTvPage.tsx]: Yes we can. _No. I'm not Obama_ 😂 
We now have a test to check this [here](https://github.com/bbc/simorgh/pull/11928/files#diff-3392579f3c5412875131f2725913fc36918dade93cb4affa1cc094570a63f7a3R224).

File: cypress/e2e/pages/onDemandTV/tests.js (hunk 12 above)
Comment (alex-magana) [reply cypress/e2e/pages/onDemandTV/tests.js:19]: This is now in place. We've updated [getEpisodeAvailability](https://github.com/bbc/simorgh/pull/11928/files#diff-be0dc0c9e10b2f5281a37e5c0dfc2a0999110ecd8e3fc0c7b4d0125611108e5aR60) to traverse the
schema we receive from `getInitialData`.
`getEpisodeAvailability` will replace [isAvailable](https://github.com/bbc/simorgh/pull/11928/files#diff-be0dc0c9e10b2f5281a37e5c0dfc2a0999110ecd8e3fc0c7b4d0125611108e5aR35) as part of the On Demand Audio migration where it's still in use.

File: cypress/e2e/pages/onDemandTV/tests.js (hunk 13 above)
Comment (alex-magana) [reply cypress/e2e/pages/onDemandTV/tests.js]: This assertion has been added [here](https://github.com/bbc/simorgh/pull/11928/files#diff-5babdcd98a0017973821535dce88c8916435f66e93ae8287287dadf218f0f8e2R33) enabled by [videoPlaceholderImageUrl](https://github.com/bbc/simorgh/pull/11928/files#diff-be0dc0c9e10b2f5281a37e5c0dfc2a0999110ecd8e3fc0c7b4d0125611108e5aR60).
